R/utils-code-dependency.R

```diff
@@ -0,0 +1,500 @@
+#' new_code_dep <- code_dependency(parse(text = code2), envir = d)
+#' bind_code_dependency(old_code_dep, new_code_dep)
+#'
+#'
+#' q1 <- teal.code:::new_qenv()
+#' q1@code_dependency
+#'
+#' q2 <- teal.code::eval_code(q1, code = code)
+#' q2@code
+#' q2@code_dependency
+#'
+#' q3 <- teal.code::eval_code(q2, code = code2)
+#' q3@code
+#' q3@code_dependency
+#'
+#' q4 <- teal.code::eval_code(q3, code = code3)
+#' q4@code
+#' q4@code_dependency
+#'
+#' get_code(q2, deparse = FALSE, names = "ADLB")
+#' get_code(q3, deparse = FALSE, names = "ADLB")
+#' get_code(q4, deparse = FALSE, names = "ADLB")
+#' get_code(q4, deparse = FALSE, names = "var_labels")
+#' get_code(q4, deparse = FALSE, names = "ADSL")
+#' get_code(q4, deparse = FALSE, names = c("ADSL", "ADS", "C"))
+#' get_code(q4, deparse = FALSE, names = c("var_labels", "ADSL"))
+#' get_code(q4)
+#' }
+#'
+#' @keywords internal
+code_dependency <- function(parsed_code, envir = new.env()) {
+  if (identical(parsed_code, expression()) || is.null(attr(parsed_code, "srcref"))) {
+    return(list())
+  }
+
+  pd <- getParseData(parsed_code)
+
+  calls_pd <- lapply(pd[pd$parent == 0, "id"], get_children, pd = pd)
+
+  object_names <- ls(envir)
+
+  occurrence <- lapply(lapply(object_names, detect_symbol, pd = calls_pd), which)
+
+  names(occurrence) <- object_names
+
+  cooccurrence <- lapply(
+    calls_pd,
+    function(x) {
+      sym_cond <- which(x$token == "SYMBOL" & x$text %in% object_names)
+      if (length(sym_cond) >= 2) {
+        ass_cond <- grep("ASSIGN", x$token)
+        text <- unique(x[sort(c(sym_cond, ass_cond)), "text"])
+
+        if (text[1] == "->") {
+          rev(text[-1])
+        } else {
+          text[-1]
+        }
+      }
+    }
+  )
+
+  side_effects <- grep("@effect", pd[pd$token == "COMMENT", "text"], value = TRUE)
+  check_effects <-
+    if (length(side_effects) > 0) {
+      affected <-
+        setdiff(
+          unlist(
+            strsplit(
+              unlist(
+                lapply(
+                  strsplit(
+                    side_effects,
+                    split = "@effect",
+                    fixed = TRUE
+                  ),
+                  function(x) x[-1]
+                )
+              ),
+              split = " ",
+              fixed = TRUE
+            )
+          ),
+          ""
+        )
+      unique(c(object_names, affected))
+    } else {
+      object_names
+    }
+
+  effects <- lapply(check_effects, return_code_for_effects, pd = calls_pd, occur = occurrence, cooccur = cooccurrence)
+  names(effects) <- check_effects
+
+  list(
+    occurrence = occurrence,
+    cooccurrence = cooccurrence,
+    effects = effects
+  )
+}
+
+#' @title Get children calls within `getParseData()` object
+#' @param pd `list` of data.frames of results of `utils::getParseData()` trimmed to unique `parsed_code` calls
+#' @param parent parent id in `utils::getParseData()`
+#' @return Row `binded` `utils::getParseData()` of all calls.
```

Comment:
`pd` supposted to be a list of `data.frame`s as each element of list corresponds to a separate call. Will double check as it was some time since I wrote the code